libfwupdplugin/fu-efi-firmware-volume.c

```diff
@@ -0,0 +1,386 @@
+#define FU_EFI_FIRMWARE_VOLUME_OFFSET_RESERVED			0x36
+#define FU_EFI_FIRMWARE_VOLUME_OFFSET_REVISION			0x37
+#define FU_EFI_FIRMWARE_VOLUME_OFFSET_BLOCK_MAP			0x38
+#define FU_EFI_FIRMWARE_VOLUME_SIZE				0x40
+
+static gboolean
+fu_efi_firmware_volume_parse_one (FuEfiFirmwareVolume *self,
+				  GBytes *fw,
+				  gsize *offset_val,
+				  guint32 img_cnt,
+				  FwupdInstallFlags flags,
+				  GError **error)
+{
+	fwupd_guid_t guid = { 0x0 };
+	gsize blockmap_sz = 0;
+	gsize bufsz = 0;
+	gsize offset = *offset_val;
+	guint16 checksum = 0;
+	guint16 ext_hdr = 0;
+	guint16 hdr_length = 0;
+	guint32 attrs = 0;
+	guint32 sig = 0;
+	guint64 fv_length = 0;
+	guint8 revision = 0;
+	const guint8 *buf = g_bytes_get_data (fw, &bufsz);
+	g_autofree gchar *guid_str = NULL;
+	g_autoptr(FuFirmwareImage) img = NULL;
+	g_autoptr(GBytes) blob = NULL;
+
+	/* check if this is _FIT */
+	if (!fu_common_read_uint32_safe (buf, bufsz, offset,
+					 &sig, G_LITTLE_ENDIAN, error)) {
+		g_prefix_error (error, "failed to read start signature: ");
+		return FALSE;
+	}
+	if (sig == FU_EFI_FIT_SIGNATURE) {
+		*offset_val += 0x150000;
+		return TRUE;
+	}
+	if (!fu_common_read_uint32_safe (buf, bufsz,
+					 offset + FU_EFI_FIRMWARE_VOLUME_OFFSET_SIGNATURE,
+					 &sig, G_LITTLE_ENDIAN, error)) {
+		g_prefix_error (error, "failed to read signature: ");
+		return FALSE;
+	}
+	if (sig != FU_EFI_FIRMWARE_VOLUME_SIGNATURE) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "signature invalid, got 0x%x, expected 0x%x",
+			     sig, (guint) FU_EFI_FIRMWARE_VOLUME_SIGNATURE);
+		return FALSE;
+	}
+
+	/* check reserved section */
+	for (guint i = 0; i < 0x10; i++) {
+		if (buf[offset + FU_EFI_FIRMWARE_VOLUME_OFFSET_ZERO_VECTOR + i] != 0x0) {
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "reserved section invalid @0x%x",
+				     (guint) offset + FU_EFI_FIRMWARE_VOLUME_OFFSET_ZERO_VECTOR + i);
+			return FALSE;
+		}
+	}
+
+	/* guid */
+	if (!fu_memcpy_safe ((guint8 *) &guid, sizeof(guid), 0x0,		/* dst */
+			     buf, bufsz, offset + FU_EFI_FIRMWARE_VOLUME_OFFSET_GUID, 	/* src */
+			     sizeof(guid), error)) {
+		g_prefix_error (error, "failed to read GUID: ");
+		return FALSE;
+	}
+	guid_str = fwupd_guid_to_string (&guid, FWUPD_GUID_FLAG_MIXED_ENDIAN);
+	g_debug ("volume GUID: %s [%s]", guid_str, fu_efi_guid_to_name (guid_str));
+
+	/* length */
+	if (!fu_common_read_uint64_safe (buf, bufsz,
+					 offset + FU_EFI_FIRMWARE_VOLUME_OFFSET_LENGTH,
+					 &fv_length, G_LITTLE_ENDIAN, error)) {
+		g_prefix_error (error, "failed to read length: ");
+		return FALSE;
+	}
+	if (fv_length == 0x0) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "invalid volume length");
+		return FALSE;
+	}
+	if (!fu_common_read_uint32_safe (buf, bufsz,
+					 offset + FU_EFI_FIRMWARE_VOLUME_OFFSET_ATTRS,
+					 &attrs, G_LITTLE_ENDIAN, error))
+		return FALSE;
```

Comment:
don't you want to prefix the error here?

Reply:
Agree, fixed.